File: insert_interval.cpp
#include "header.h"

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
    vector<Interval> res;
    int r=0;
    while(r<intervals.size() && newInterval.start>intervals[r].end) r++;
    if(r==intervals.size()){
        intervals.push_back(newInterval);
        return intervals;
    }
    intervals.insert(intervals.begin()+r, newInterval);
    while(r+1<intervals.size()){
        if(intervals[r+1].start>intervals[r].end) break;
        intervals[r].start = std::min(intervals[r].start,intervals[r+1].start);
        intervals[r].end = std::max(intervals[r].end,intervals[r+1].end);
        intervals.erase(intervals.begin()+r+1);
    }
    return intervals;
}

int main(){
    Interval i0(1, 2);
    Interval i1(3, 5);
    Interval i2(6, 7);
    Interval i3(8, 10);
    Interval i4(12, 16);
    Interval i5(4,9);
    vector<Interval> intervals = {i0, i1, i2, i3, i4};
    vector<Interval> res = insert(intervals, i5);
    for(int i=0; i<res.size(); i++){
        cout<<res[i].start<<" "<<res[i].end<<endl;
    }
    return 0;
}

File: master/insert_interval.py

# Definition for an interval.
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e
    def __repr__(self):
        return "start: {0}, end: {1}\n".format(self.start,self.end)

class Solution:
    # @param intervals, a list of Intervals
    # @param newInterval, a Interval
    # @return a list of Interval
    def insert(self, intervals, newInterval):
        i = 0
        while i<len(intervals) and intervals[i].end<newInterval.start:
            i+=1
        if i == len(intervals):
            intervals.append(newInterval)
            return intervals
        intervals.insert(i,newInterval)
        while i+1<len(intervals):
            if intervals[i].end<intervals[i+1].start:
                break
            intervals[i].start = min(intervals[i].start,intervals[i+1].start)
            intervals[i].end = max(intervals[i].end, intervals[i+1].end)
            intervals.pop(i+1)
        return intervals

if __name__ == "__main__":
    i0 = Interval(1,2)
    i1 = Interval(3,5)
    i2 = Interval(6,7)
    i3 = Interval(8,10)
    i4 = Interval(12,16)
    i5 = Interval(4,9)
    print Solution().insert([i0, i1, i2, i3, i4], i5)
